File: frontend/src/store/trading.ts
import { create } from 'zustand';
import api from '@/lib/api';

export interface Position {
  id: string;
  stock_code: string;
  stock_name: string | null;
  quantity: number;
  avg_cost_price: number;
  current_price: number | null;
  unrealized_pnl: number | null;
  realized_pnl: number;
}

export interface Order {
  id: string;
  stock_code: string;
  stock_name: string | null;
  side: string;
  order_type: string;
  quantity: number;
  limit_price: number | null;
  filled_quantity: number;
  avg_fill_price: number | null;
  status: string;
  submitted_at: string | null;
  filled_at: string | null;
  created_at: string;
}

export interface PortfolioStats {
  portfolio_value: number;
  total_invested: number;
  unrealized_pnl: number;
  realized_pnl: number;
  total_pnl: number;
  total_pnl_pct: number;
  position_count: number;
  trade_stats: {
    total_trades: number;
    win_rate: number;
    profit_factor: number;
    winning_trades: number;
    losing_trades: number;
  };
}

interface TradingState {
  positions: Position[];
  orders: Order[];
  portfolioStats: PortfolioStats | null;
  selectedStock: string | null;
  isLoading: boolean;
  lastUpdated: number | null;
  fetchPositions: () => Promise<void>;
  fetchOrders: () => Promise<void>;
  fetchPortfolioStats: () => Promise<void>;
  setSelectedStock: (code: string | null) => void;
  updatePositionPrice: (stockCode: string, price: number) => void;
}

export const useTradingStore = create<TradingState>((set, get) => ({
  positions: [],
  orders: [],
  portfolioStats: null,
  selectedStock: null,
  isLoading: false,
  lastUpdated: null,

  fetchPositions: async () => {
    set({ isLoading: true });
    try {
      const { data } = await api.get('/trading/positions');
      set({ positions: data, lastUpdated: Date.now() });
    } finally {
      set({ isLoading: false });
    }
  },

  fetchOrders: async () => {
    try {
      const { data } = await api.get('/trading/orders');
      set({ orders: data });
    } catch {
      // handle error
    }
  },

  fetchPortfolioStats: async () => {
    try {
      const { data } = await api.get('/trading/portfolio/analytics');
      set({ portfolioStats: data });
    } catch {
      // handle error
    }
  },

  setSelectedStock: (code) => set({ selectedStock: code }),

  updatePositionPrice: (stockCode, price) => {
    const { positions } = get();
    const updated = positions.map((p) => {
      if (p.stock_code !== stockCode) return p;
      const unrealized = (price - p.avg_cost_price) * p.quantity;
      return { ...p, current_price: price, unrealized_pnl: unrealized };
    });
    set({ positions: updated, lastUpdated: Date.now() });
  },
}));

File: backend/app/schemas/trading.py
from pydantic import BaseModel
from datetime import datetime
from decimal import Decimal


class OrderCreate(BaseModel):
    stock_code: str
    stock_name: str | None = None
    side: str  # 'buy' or 'sell'
    order_type: str = "market"  # 'market', 'limit'
    quantity: int
    limit_price: float | None = None
    strategy_id: str | None = None


class OrderResponse(BaseModel):
    id: str
    stock_code: str
    stock_name: str | None
    side: str
    order_type: str
    quantity: int
    limit_price: float | None
    filled_quantity: int
    avg_fill_price: float | None
    status: str
    submitted_at: datetime | None
    filled_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class PositionResponse(BaseModel):
    id: str
    stock_code: str
    stock_name: str | None
    quantity: int
    avg_cost_price: float
    current_price: float | None
    unrealized_pnl: float | None
    realized_pnl: float

    model_config = {"from_attributes": True}


class BalanceResponse(BaseModel):
    total_balance: float
    available_cash: float
    invested_amount: float
    total_pnl: float
    total_pnl_percent: float
